src/grpc.cc

```diff
@@ -170,9 +170,87 @@ static grpc::Status marshal_value_list(const value_list_t *vl,
   msg->set_allocated_time(new google::protobuf::Timestamp(t));
   msg->set_allocated_interval(new google::protobuf::Duration(d));
 
+  msg->clear_meta_data();
+  if (vl->meta != nullptr) {
+    char **meta_data_keys = nullptr;
+    int meta_data_keys_len = meta_data_toc(vl->meta, &meta_data_keys);
+    if (meta_data_keys_len < 0) {
+      return grpc::Status(grpc::StatusCode::INTERNAL,
+                          grpc::string("error getting metadata keys"));
+    }
+
+    for (int i = 0; i < meta_data_keys_len; i++) {
+      char *key = meta_data_keys[i];
+      int md_type = meta_data_type(vl->meta, key);
+
+      collectd::types::MetadataValue md_value;
+      md_value.Clear();
+
+      switch (md_type) {
+      case MD_TYPE_STRING:
+        char *md_string;
+        if (meta_data_get_string(vl->meta, key, &md_string) != 0 || md_string == nullptr) {
+          strarray_free(meta_data_keys, meta_data_keys_len);
+          return grpc::Status(grpc::StatusCode::INTERNAL,
+                            grpc::string("missing metadata"));
+        }
+        md_value.set_string_value(md_string);
+        free(md_string);
+        break;
+      case MD_TYPE_SIGNED_INT:
+        int64_t int64_value;
+        if (meta_data_get_signed_int(vl->meta, key, &int64_value) != 0) {
+          strarray_free(meta_data_keys, meta_data_keys_len);
+          return grpc::Status(grpc::StatusCode::INTERNAL,
+                            grpc::string("missing metadata"));
+        }
+        md_value.set_int64_value(int64_value);
+        break;
+      case MD_TYPE_UNSIGNED_INT:
+        uint64_t uint64_value;
+        if (meta_data_get_unsigned_int(vl->meta, key, &uint64_value) != 0) {
+          strarray_free(meta_data_keys, meta_data_keys_len);
+          return grpc::Status(grpc::StatusCode::INTERNAL,
+                            grpc::string("missing metadata"));
+        }
+        md_value.set_uint64_value(uint64_value);
+        break;
+      case MD_TYPE_DOUBLE:
+        double double_value;
+        if (meta_data_get_double(vl->meta, key, &double_value) != 0) {
+          strarray_free(meta_data_keys, meta_data_keys_len);
+          return grpc::Status(grpc::StatusCode::INTERNAL,
+                            grpc::string("missing metadata"));
+        }
+        md_value.set_double_value(double_value);
+        break;
+      case MD_TYPE_BOOLEAN:
+        bool bool_value;
+        if (meta_data_get_boolean(vl->meta, key, &bool_value) != 0) {
+          strarray_free(meta_data_keys, meta_data_keys_len);
+          return grpc::Status(grpc::StatusCode::INTERNAL,
+                            grpc::string("missing metadata"));
+        }
+        md_value.set_bool_value(bool_value);
+        break;
+      default:
+        strarray_free(meta_data_keys, meta_data_keys_len);
+        return grpc::Status(grpc::StatusCode::INTERNAL,
```

Comment:
This patch adds an error message for this case when handling metric values. For symmetry I think we should have an error message here, too.